### DISCUSSION TITLE
NEW File Manager for ESP32 runs directly on device in nonblocking mode

### DISCUSSION THREAD

**Discussion by mispacek** [micropython/micropython#15475] [ESP32]
### **Online file manager for ESP32 micropython**

https://github.com/mispacek/ESP-File_manager

**Manage your files in ESP over WIFI**

Supported functions
- Upload files (Drag and drop to browser window is supported, you can drop multiple files)
- Copy multiple files to another directory
- Move multiple files to another directory
- Download multiple files to your PC
- New folder creation
- Delete files on ESP
- Mobile phone optimized

![Preview](https://github.com/user-attachments/assets/b722a89c-3ca3-424a-98cb-9275466276c3)

**Comment by walklinewang** [comment]
I tested this file manager on my ESP32C3 board and it's amazing how smooth it is, it's also the first time I've seen such a small WebServer.

I will keep watching this repo, and thanks to your sharing.

![picture](https://github.com/user-attachments/assets/ee0c04fb-4d2e-4482-a908-a212192cb06f)

**Reply by davefes** [reply]
Ran on a ESP32-WROOM-32 using Ubuntu running Firefox.  I currently use `rshell` and see how this makes moving files around a lot easier.  

Interested in how people would integrate this, so that you could run your program on the device and for debugging.

**Reply by walklinewang** [reply]
I made an attempt to use the file manager as a package.

![image](https://github.com/user-attachments/assets/cca12381-6599-42ef-9f18-5756f5f1be72)

The `__init__.py` is a copy of the `main.py` file, with the following differences:

```python
# using related import to replace original
from .web_server import WebServer
from .filemanager import *

# changing web_folder
webserver = WebServer(web_folder='./filemanager/www', port=80)

# in the repl, import and start file manager
import filemanager
```

The only problem is how to upload 18 files to your board...

**Reply by mispacek** [reply]
Thanks for your great work. It can be useful, to use it as module.

> The only problem is how to upload 18 files to your board...

I think it cannot be too hard. I develop system for installing my other project to ESP board. Its combination of Wget and Tar unpack

This is two files with related code, sorry for comments in Czech, its local project for kids education.
Installer with tar decompress: https://pastebin.com/b4VzeGtY
Wget for downloading tar archive: https://pastebin.com/Sftv3kpz

Thonny is usefull tool too. It can upload files very well.

I will be glad for further information as the project continues
It would be great to work together on further development.

**Reply by walklinewang** [reply]
Thanks for the ideas, the Tar file is indeed a good choice.

There are many ways to upload these files, by problem I mean for people __willing__ to try to use it, as you say, Thonny also makes it easy to upload files, including deleting, renaming and even editing them on-board, and your file manager overlaps with it in terms of usage, and uploading 18 more files in that context, pessimistically, many people won't be willing to try it, I've also have written a couple of efficiency tools, so I know this all too well.

I'm going to try what you've said, firstly packaging the files into a Tar file, then figuring out how to compile them into the firmware, and then extracting them to the user's partition to use as a module on first time.

Translated with DeepL.com (free version)   -- Yes, keeping this statement I just want to say that English is really hard -_-|||

**Reply by mispacek** [reply]
File manager is only part of my ESP IDE. I need file manager for importing libraries, exporting data and deleting of old files. 
But for today, whole project is only in czech language, now I working on updates and multilanguage support
https://youtube.com/shorts/jPLrap-1XfI

https://github.com/mispacek/ESP-IDE-ESP32C3
https://iotserver.cz/    you can use google translate for translating to english

I didn't expect anyone to use the file manager as a standalone project, but it's very suitable for integration into something more complex

It would be good to optimize the python part of the filemanager to be less RAM intensive.
